Management: Update docs for new device statuses and update status (#3078)
balena-io/docs: pages/learn/manage/device-statuses.md

> 

Reviewer: it feels a bit clunky that the above image is between "the following statuses" and the table of statuses

Reviewer 2: agreed - I also feel the image needs a description as to what it's showing? it took me a while to connect the colourful bar as the statuses.

Author: Agreed, I think it's actually clearer if I remove that "fleet" summary image, it doesn't add anything and is fairly unrelated to device status

> | Status                                       | Description                                                                                                                                                                      |

Reviewer 2: I think we need to have a CTA that links to the blog post on this page - as that details each of the direct path changes e.g. online to operational. Perhaps detailing the date this changed

e.g. [date] We have recently made changes to the naming of device statuses. You can find out more details here.

> | Status                                       | Description                                                                                                                                                                      |

Reviewer 2: I think we need to have a CTA that links to the blog post on this page - as that details each of the direct path changes e.g. online to operational. Perhaps detailing the date this changed

e.g. [date] We have recently made changes to the naming of device statuses. You can find out more details here.

|----------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **Operational**                              | The device is functioning normally and communicating with the cloud within the expected interval. In the standard configuration, this means both [cloudlink][cloudlink] and the device heartbeat API are working. If [cloudlink is disabled][bandwidth-reduction], the device will still appear `Operational` if the heartbeat API is functional.   |

Reviewer 2: maybe 'expected time interval'?

> | **Operational**                              | The device is functioning normally and communicating with the cloud within the expected interval. In the standard configuration, this means both [cloudlink][cloudlink] and the device heartbeat API are working. If [cloudlink is disabled][bandwidth-reduction], the device will still appear `Operational` if the heartbeat API is functional.   |

Reviewer 2: maybe 'expected time interval'?

Reviewer 3: +1

| **Reduced Functionality**<sup>1</sup>        | The device is connected to the backend but is experiencing limited functionality due to either **missing recent Heartbeat API communications** or **lack of connection to cloudlink**.                                                                                       |

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: You've linked to [this page](https://d89e6939.balenacloud-docs.pages.dev/learn/manage/configuration/#variable-list) here but from a UX perspective, the API polling interval or poll interval isn't mentioned here? It's confusing from UX

Reviewer 2: You've also linked to docs here when mentioning the interval, but not under the mention of interval under 'operational' - is this intentional?

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: Are we also happy using the term 'offline' here?

Reviewer 2: 'the cloud' - should it be 'balenaCloud'?

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: Are we happy using 'offline' here?

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: You've linked to [this page](https://d89e6939.balenacloud-docs.pages.dev/learn/manage/configuration/#variable-list) here but from a UX perspective, the API polling interval or poll interval isn't mentioned here? It's confusing from UX

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: You've also linked to docs here when mentioning the interval, but not under the mention of interval under 'operational' - is this intentional?

> | **Disconnected**                                  | The device is not connected to cloudlink and has not communicated via the Heartbeat API for a duration exceeding the [API polling interval][poll-interval]. This status indicates that the device is completely offline and unable to send or receive any data from the cloud.                                                                                                                                                           |

Reviewer 2: Should it be 'balenaCloud' instead of just 'cloud'?

| **Configuring**                              | The device is applying OS configuration. For example, an OS upgrade.                                                                                                                                         |
| **Updating**                                 | The device is updating to a new App release. In future versions this will also reflect OS updates.                                                                                                                                        |

> | **Post Provisioning**                        | The device has been [provisioned][device-provisioning] but has not yet come online.                                                                                              |

Reviewer 2: Should we be saying 'online' here?

> | **Post Provisioning**                        | The device has been [provisioned][device-provisioning] but has not yet come online.                                                                                              |

Reviewer 2: Should we be saying 'online' here?

| **Inactive**                                 | The device has been [deactivated][deactivated] or has been [preregistered][preregistered] but has not yet connected to the {{ $names.cloud.lower }} API.                         |
| **Frozen**                                   | The device has been frozen because it's outside the organization's allowance, or is in a paid [fleet type][fleet type] on a free tier organization.                              |

> ## Device Connectivty states

Reviewer: typo

```suggestion
## Device Connectivity states
```

Reviewer 2: I think we need to have a CTA that links to the blog post on this page - as that details each of the direct path changes _e.g. online to operational._  Perhaps detailing the date this changed

e.g. [date] We have recently made changes to the naming of device statuses. You can find out more details here.

> The balena device uses two mechanisms to maintain communication with balenaCloud, namely a periodic HTTP API heartbeat and a VPN based tunnel called cloudLink. The state of both of these links can be seen in the device listing.

Reviewer: different capitalisation

```suggestion
The balena device uses two mechanisms to maintain communication with balenaCloud, namely a periodic HTTP API heartbeat and a VPN based tunnel called cloudlink. The state of both of these links can be seen in the device listing.
```

> ![Device connectivity indicators](/img/common/main_dashboard/device_status.webp)

Reviewer 2: Is it overkill to explain why a user would see '-' instead of a status? I know we've added a note of 'data for these states will only get populated for devices with supervisor 16.7.0 or later' but should we have a catch all as to why this shows? e.g. there's no data to populate this section

Author: That feels a bit overkill

> <sup>1</sup> `Reduced Functionality` indicates that the device is experiencing limited connectivity. This can be due to either **missing recent Heartbeat API communications** or **lack of connection to cloudlink**. The boolean columns for these states are as follows:

Reviewer: The superscript 1 reference was hard to find - I expected it to be directly below the table or at the very end of the entire document, or perhaps clickable?

> - **Cloudlink Available, but heartbeat not**: The device is unable to communicate with the {{ $names.cloud.lower }} API. A device with cloudlink only is not able to apply any new changes made such as deploying new releases, applying service configuration values, or switching to local mode. However, it is accessible via SSH or the web terminal. Performing an action such as rebooting or restarting containers might work, but most likely will not. This is because the device loses its Heartbeat if it's not communicating with the API, which is usually when the Supervisor on the device is not running or crashing. Since we have cloudlink tunnel access, we can SSH into the device and investigate further.

Reviewer: found this to read with a lack of confidence - 'might work' is probably not something to put in documentation.

"we have cloudlink tunnel access..." might need rephrasing too. 

```suggestion
- **Cloudlink Available, but heartbeat not**: A missing heartbeat means that the device is unable to communicate with the {{ $names.cloud.lower }} API. A device in this state is not able to apply any new changes made such as deploying new releases, applying service configuration values, or switching to local mode, due to a reliance on API connectivity for this functionality. A missing heartbeat is symptomatic of a problem with the device connectivity or Supervisor, meaning actions such as rebooting or restarting containers are likely to be unavailable. Device debugging is possible via the terminal in this state due to the availability of cloudlink which allows an on-demand connection to the device to be made.
```

Reviewer 2: Why is 'Available' capitalised in both bullet points?

> - **Cloudlink Available, but heartbeat not**: The device is unable to communicate with the {{ $names.cloud.lower }} API. A device with cloudlink only is not able to apply any new changes made such as deploying new releases, applying service configuration values, or switching to local mode. However, it is accessible via SSH or the web terminal. Performing an action such as rebooting or restarting containers might work, but most likely will not. This is because the device loses its Heartbeat if it's not communicating with the API, which is usually when the Supervisor on the device is not running or crashing. Since we have cloudlink tunnel access, we can SSH into the device and investigate further.

Reviewer 2: Why is 'Available' capitalised in both bullet points?

> - **Heartbeat Available, but cloudlink not**: The device is unable to connect to cloudlink (e.g. a firewall is blocking VPN traffic), and won't be able to provide remote SSH connections. A device with Heartbeat Only status has internet connectivity and can poll the cloud for new updates to apply. However, performing any actions which do not change the target state or attempting to access the device via SSH will error, as these actions are proxied via the cloudlink tunnel connection. Actions that do not change the target state include purging data, restarting, rebooting, or shutting down.

Reviewer: ```suggestion
- **Heartbeat Available, but cloudlink not**: The device is unable to connect to cloudlink (e.g. a firewall is blocking VPN traffic), and terminal access to the device is thus unavailable. A device with Heartbeat Only status has internet connectivity, is able to reach the API, and can poll the cloud for new updates to apply, but actions that take immediate effect, such as purging data, restarting services, rebooting or shutting down, will not be available as they are performed via cloudlink.
```

> __Note:__ If the device is powered off or loses all network connectivity, the `heartbeat` indictor show as connected (green) until the last target state fetch exceeds the device [API polling interval][poll-interval], at which time it will be marked as `Disconnected`.

Reviewer: ```suggestion
__Note:__ If the device is powered off or loses all network connectivity, the `heartbeat` indicator will show as connected (green) until enough time has passed to cover the device [API polling interval][poll-interval], at which time it will be marked as `Disconnected`.
```

> __Note:__ If the device is powered off or loses all network connectivity, the `heartbeat` indictor show as connected (green) until the last target state fetch exceeds the device [API polling interval][poll-interval], at which time it will be marked as `Disconnected`.

Reviewer 2: same as above: this is linked to [this page](https://d89e6939.balenacloud-docs.pages.dev/learn/manage/configuration/#variable-list) but from a UX perspective, the API polling interval or poll interval isn't mentioned here? It's confusing from UX

> __Note:__ If the device is powered off or loses all network connectivity, the `heartbeat` indictor show as connected (green) until the last target state fetch exceeds the device [API polling interval][poll-interval], at which time it will be marked as `Disconnected`.

Author: I think we need to create a page or section somewhere else explaining what the poll interval is. I will leave that for a separate PR. Maybe @vipulgupta2048 knows where it would be best to fit that in?

## Debugging Device Status

If you find your device to be displaying a status which is unclear even with the above notes, visit the [device debugging masterclass][debugging-masterclass] for more information.

> 

Reviewer 2: The padding here is a bit weird? It looks like 'Update Status' is within the 'Debugging Device Status' section. I'd suggest adding a bigger gap here.

# Update Status

The `Update status` and `Update status duration` fields on devices represent the state of the current application update to target state. 

> __Note:__ Data for these states will only get populated for devices with supervisor 16.7.0 or later.

Reviewer 2: Do we want to link to the announcement blog post here?

## Update statuses

| Status                                       | Description                                                                                                                                                                      |
|----------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Done**                              | The update, restart or configuration change is complete and the device is in its target state.                                          |
| **Downloading**                       | The supervisor is downloading one or more container images associated with the target release                                           |

> | **Downloaded**                        | All container images in the target release have been downloaded. **Note:** If the device holds an update lock, it will stop in this state until the lock is released (depending on the [update strategy](https://docs.balena.io/learn/deploy/release-strategy/update-strategies/#controlling-the-update-strategy)). |

Reviewer 2: Maybe put the note on a separate line here

Author: unfortunately it doesn't seem possible to do within markdown. I'm sure there is a way to do it with html in markdown but it seems non-trivial :/

| **Applying Changes**                  | The supervisor is performing actions to install the new release, i.e. killing old containers and making volumes/network/container changes.    |

> | **Rejected**                          | The target release has required services that do not meet the [necessary requirements](https://docs.balena.io/learn/develop/container-contracts/#container-contracts) to run. |

Reviewer 2: It's a minor point, but the lack of the phrase 'necessary requirements' mentioned in the linked documentation isn't brilliant UX. I appreciate requirements are mentioned lots though, so this isn't as big a thing as the polling interval comment.

| **Aborted**                           | The supervisor has given up on trying to install a new release. There are currently no conditions under which the supervisor returns this state, but future work could allow the supervisor to abort an update, for instance, after a number of failed download tries.    |

## Update Status Duration

The `Update Status Duration` field is calculated based on the timestamp of the most recent update event. When querying this data from the API, you will receive a timestamp instead of a duration.

> It is important to note that when the application starts up after a reboot or an app restart, this behavior is indistinguishable from an app installation. Consequently, the `Update Status Duration` timestamp will be refreshed whenever a reboot or restart occurs.

Reviewer: This is potentially confusing - does this mean most devices in a stable state will say Done with an update status that's essentially the device or service uptime? I would expect (with this wording) for it to show the time since the update completed. Unsure what to suggest as an alternative at the moment 🤔

Author: @chrisys yeah unfortunately that is the case. If you rebooted a device, its `update status duration` would be equal to the "uptime" (minus supervisor boot time). I spoke to felipe about this but there isn't a simple way for the supervisor currently to distinguish between starting a composition of containers because of an update or because of restart/reboot. We will need to do a fairly large restructure of the supervisor for that :/

## Statuses and OS upgrades

> The `Update status` currently focuses on application updates, and for the time being OS updates are still reflected as part of the device status. When a device is performing an OS update, the status will change to `Configuring` with additional details in parentheses, such as `Configuring (Update successful, rebooting)`. While filtering or searching based on the data in parentheses is not yet available, we are working towards unifying all update paths under the `Update Status` mechanism used by applications. This will make it even easier to track and manage updates in the future.

Reviewer 2: Blog post link here? Particularly when referencing future work
